File: 프로그래머스/Level 1/달리기 경주.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/178871
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

vector<string> solution(vector<string> players, vector<string> callings) {
    vector<string> answer;
    unordered_map<string, int> playersMap;
    
    for (int i = 0; i < players.size(); ++i)
    {
        playersMap[players[i]] = i;
    }
    
    for (string c : callings)
    {
        // 현재 불린 선수의 앞에 선수
        string prev = players[playersMap[c]-1];
        
        playersMap[c] -= 1;
        playersMap[prev] += 1;
        
        string tmp =  players[playersMap[c]];
        players[playersMap[c]] = players[playersMap[prev]];
        players[playersMap[prev]] = tmp;
    }
    
    answer = players;
    return answer;
}

File: 프로그래머스/Level 1/달리기 경주.swift
// https://school.programmers.co.kr/learn/courses/30/lessons/178871
import Foundation

func solution(_ players:[String], _ callings:[String]) -> [String] {
    var answer: [String] = players
    var rank: [String: Int] = [:]
    
    for player in players.enumerated() {
        rank[player.1] = player.0
    }
    
    //print(rank)
    
    for calling in callings {
        let cur = rank[calling]!
        let prev = cur - 1
        rank[calling]? -= 1
        rank[answer[prev]]? += 1
        answer.swapAt(prev, cur)

    }
    //print(rank)
    return answer
}